test/e2e/storage/csi_mock_volume.go

```diff
@@ -892,7 +892,11 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
 			loop:
 				for {
 					select {
-					case event := <-pvcWatch.ResultChan():
+					case event, ok := <-pvcWatch.ResultChan():
+						if !ok {
+							framework.Failf("PVC watch ended prematurely")
```

Comment:
How often does this happen? I suppose it is caused by load on the apiserver?

I'm wondering whether it would be better to skip the test when this happens. As proposed here, we still have a flaky test failure, it just fails without flooding the logs.

Reply:
@pohly your call, I'm happy to skip or fail. I've seen it a couple of times on our OpenShift tests on GCP. It may be related to leader elections on the apiserver (which can happen at any time), which can cause short blips where the watchers close their existing watch and relist everything when the apiserver comes back.

Reply:
I prefer skipping the test in this case.

Reply:
Watches can time out at ANY time, and both etcd and kube-apiserver are allowed to do this.

No one should ever read from a watch stream and assume how long that watch is open.  We reserve the right to close the watch at any point from kube-apiserver.

So please, fix the test to not assume things that aren't true :)

Reply:
> fix this to retry the watch (single shot watch is not safe)

I think that restarting the watch is not going to work: what the test tries to verify is that rescheduling happened. The only indication of that through the API is that the pod annotation gets set and later removed. If the watch is closed while the annotation is set and later gets restarted when it is set again, the test misses the "annotation was reset" part and fails incorrectly.

I see several options:
- stop using watches and instead rely on log parsing to detect whether rescheduling happened (not sure though whether there is anything in the logs about it)
- skip then when the watch is closed
- remove the test

I'm leaning towards skipping because log parsing may also be problematic and a test that usually (I know, that's an assumption) works is better than no test.

Reply:
Perhaps the test could use a SharedInformer on top of the watch which would handle restarts correctly via the Reflector, and the DeltaFifo would ensure that you have correct updates to the object itself. You don't care about actual watch events, you only care about changes to the PVC object itself, right?  SharedInformer would let you concentrate on just the object and not the watch.

Reply:
But does SharedInformer guarantee that I see all updates? My understanding of it is that it is possible to miss updates, i.e. the only guarantee is that the callbacks eventually get the final state of each object.

Reply:
> But does SharedInformer guarantee that I see all updates? My understanding of it is that it is possible to miss updates, i.e. the only guarantee is that the callbacks eventually get the final state of each object.

@pohly doesn't that indicate a flaw in the scheme of annotation removal/addition? Any client using shared informer might miss those PVC annotation changes too, if they are collapsed into a single update by DeltaFIFO and thus the annotation removal is elided.

I tried the sharedinformer approach in https://github.com/kubernetes/kubernetes/pull/93720 but don't think I'd have the time to push it forward. But it still the flaw you allude to; I think clients and the testcase need to assume they will only see the *current* object state and perhaps not any intermediate states, and only act on that (eg reconciliation).

Reply:
No, the actual usage of the annotation is not affected because it's value-triggered: if set, the external-provisioner does something. If unset, the scheduler does something. Nothing else changes the annotation, therefore it is fine to just know the current state.

The difference is that test wants to verify *how* we got to the current state. If the code and/or test is flawed, then we might reach the same observable end-state without actually having gone through the rescheduling cycle in the middle.

Reply:
I ended up just skipping the additional checks, i.e. the test will still be considered a success: https://github.com/kubernetes/kubernetes/pull/93777

@dcbw: perhaps you can review that PR?

Reply:
> If the watch is closed while the annotation is set and later gets restarted when it is set again, the test misses the "annotation was reset" part and fails incorrectly.

The watchtools helpers provide a watch interface backed by an informer which will restart a dropped watch from the last observed event, so it won’t miss any events as long as it gets an updated starting resourceVersion before compaction, which it generally should because watch bookmarks are now enabled. I’d recommend using that.